networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRangeMessageHandler.java

```diff
@@ -213,103 +201,52 @@ class RequestState {
     }
 
     SafeFuture<Optional<BlobSidecar>> loadNextBlobSidecar() {
-      // currentBlock is used to avoid querying the combinedChainDataClient for the same slot again
-      if (maybeCurrentBlock.isEmpty()) {
+      if (iterator.isEmpty()) {
         return combinedChainDataClient
-            .getBlockAtSlotExact(currentSlot.get(), headBlockRoot)
+            .getBlobSidecarKeys(startSlot, endSlot, maxRequestBlobSidecars)
             .thenCompose(
-                block -> {
-                  maybeCurrentBlock = block;
-                  final MiscHelpersDeneb miscHelpersDeneb =
-                      spec.atSlot(currentSlot.get()).miscHelpers().toVersionDeneb().orElseThrow();
-                  blobSidecarsCount = miscHelpersDeneb.getBlobSidecarsCount(block);
-                  return retrieveBlobSidecar();
+                list -> {
+                  iterator = Optional.of(list.iterator());
+                  return getNextElement(iterator.get());
                 });
       } else {
-        return retrieveBlobSidecar();
+        return getNextElement(iterator.get());
       }
     }
 
-    boolean isComplete() {
-      return currentSlot.get().isGreaterThan(maxSlot)
-          || UInt64.valueOf(sentBlobSidecars.get()).isGreaterThanOrEqualTo(maxRequestBlobSidecars);
-    }
-
-    void incrementSlotAndIndex() {
-      if (currentIndex.get().equals(getMaxBlobSidecarIndex())) {
-        currentIndex.set(UInt64.ZERO);
-        currentSlot.updateAndGet(UInt64::increment);
+    private SafeFuture<Optional<BlobSidecar>> getNextElement(
+        final Iterator<SlotAndBlockRootAndBlobIndex> iterator) {
+      if (iterator.hasNext()) {
+        return combinedChainDataClient.getBlobSidecarByKey(iterator.next());
       } else {
-        currentIndex.updateAndGet(UInt64::increment);
+        return SafeFuture.completedFuture(Optional.empty());
       }
     }
 
-    @NotNull
-    private SafeFuture<Optional<BlobSidecar>> retrieveBlobSidecar() {
-      SafeFuture<Optional<BlobSidecar>> blobSidecar =
-          maybeCurrentBlock
-              .map(SignedBeaconBlock::getRoot)
-              .map(
-                  blockRoot ->
-                      combinedChainDataClient.getBlobSidecarByBlockRootAndIndex(
-                          blockRoot, currentIndex.get()))
-              .orElse(SafeFuture.completedFuture(Optional.empty()));
-
-      refreshCurrentBlock();
-
-      return blobSidecar;
-    }
-
-    private void refreshCurrentBlock() {
-      maybeCurrentBlock.ifPresent(
-          block -> {
-            if (block.getSlot().equals(currentSlot.get())
-                && currentIndex.get().equals(getMaxBlobSidecarIndex())) {
-              maybeCurrentBlock = Optional.empty();
-            }
-          });
-    }
-
-    private UInt64 getMaxBlobSidecarIndex() {
-      return UInt64.valueOf(blobSidecarsCount).minusMinZero(UInt64.ONE);
+    boolean isComplete() {
+      return endSlot.isLessThan(startSlot)
```

Comment:
Request is `startSlot, count`, we consider request (19, 0) as legal and don't count it as error. Internally we interpret it as `startSlot, endSlot` as (19, 18) because (19, 19) should return blobs for 19th slot. I don't want penalize peers for 0 count request by the policy "forgive everything not criminal,  but don't act yourself to be punished (even if you don't punish for such things)"